esquery.js

```diff
@@ -334,38 +334,63 @@ function subjects(selector, ancestor) {
     return results;
 }
 
+/**
+* @callback TraverseVisitor
+* @param {?external:AST} node
+* @param {?external:AST} parent
+* @param {external:AST[]} ancestry
+*/
+
 /**
  * From a JS AST and a selector AST, collect all JS AST nodes that
  * match the selector.
  * @param {external:AST} ast
  * @param {?SelectorAST} selector
+ * @param {TraverseVisitor} visitor
  * @returns {external:AST[]}
  */
-function match(ast, selector) {
-    const ancestry = [], results = [];
-    if (!selector) { return results; }
+function traverse(ast, selector, visitor) {
+    if (!selector) { return; }
+    const ancestry = [];
     const altSubjects = subjects(selector);
     estraverse.traverse(ast, {
         enter (node, parent) {
             if (parent != null) { ancestry.unshift(parent); }
             if (matches(node, selector, ancestry)) {
                 if (altSubjects.length) {
                     for (let i = 0, l = altSubjects.length; i < l; ++i) {
-                        if (matches(node, altSubjects[i], ancestry)) { results.push(node); }
+                        if (matches(node, altSubjects[i], ancestry)) {
+                            visitor(node, parent, ancestry);
+                        }
                         for (let k = 0, m = ancestry.length; k < m; ++k) {
                             if (matches(ancestry[k], altSubjects[i], ancestry.slice(k + 1))) {
-                                results.push(ancestry[k]);
+                                visitor(ancestry[k], parent, ancestry);
```

Comment:
For this case, should we be passing ancestry or saving the sliced ancestry and passing that?

Reply:
I would think it should be consistent--that the `visitor` either always gets a reference or always get a copy of the ancestry, no? Also, the above snippet is only slicing part of the ancestry, and in the other instances where the `visitor` is called, it is passing in the entire ancestry.

Reply:
The slicing here isn't simply for copying though, it participates in the match. So I think the visitor would like to see the ancestry that was used for matching, right?

Reply:
Maybe so. I've added a commit, but tbh, I'm not sure (though your reasoning sgtm).

Reply:
Let's go with this for now and revise if consumers have use cases for the full ancestry.